// @storybook/sveltekit/dist/mocks/app/navigation.js
import "../../_browser-chunks/chunk-4BE7D4DS.js";

// src/mocks/app/navigation.ts
import { getContext, onMount, setContext } from "svelte";
async function goto(...args) {
  let event = new CustomEvent("storybook:goto", {
    detail: args
  });
  window.dispatchEvent(event);
}
function setAfterNavigateArgument(afterNavigateArgs) {
  setContext("after-navigate-args", afterNavigateArgs);
}
function afterNavigate(cb) {
  let argument = getContext("after-navigate-args");
  onMount(() => {
    cb && cb instanceof Function && cb(argument);
  });
}
function onNavigate() {
}
function beforeNavigate() {
}
function disableScrollHandling() {
}
async function invalidate(...args) {
  let event = new CustomEvent("storybook:invalidate", {
    detail: args
  });
  window.dispatchEvent(event);
}
async function invalidateAll() {
  let event = new CustomEvent("storybook:invalidateAll");
  window.dispatchEvent(event);
}
function preloadCode() {
}
function preloadData() {
}
async function pushState(...args) {
  let event = new CustomEvent("storybook:pushState", {
    detail: args
  });
  window.dispatchEvent(event);
}
async function replaceState(...args) {
  let event = new CustomEvent("storybook:replaceState", {
    detail: args
  });
  window.dispatchEvent(event);
}
export {
  afterNavigate,
  beforeNavigate,
  disableScrollHandling,
  goto,
  invalidate,
  invalidateAll,
  onNavigate,
  preloadCode,
  preloadData,
  pushState,
  replaceState,
  setAfterNavigateArgument
};
