// @storybook/sveltekit/dist/preset.js
import CJS_COMPAT_NODE_URL_to5drcle0z from 'node:url';
import CJS_COMPAT_NODE_PATH_to5drcle0z from 'node:path';
import CJS_COMPAT_NODE_MODULE_to5drcle0z from "node:module";

var __filename = CJS_COMPAT_NODE_URL_to5drcle0z.fileURLToPath(import.meta.url);
var __dirname = CJS_COMPAT_NODE_PATH_to5drcle0z.dirname(__filename);
var require = CJS_COMPAT_NODE_MODULE_to5drcle0z.createRequire(import.meta.url);

// ------------------------------------------------------------
// end of CJS compatibility banner, injected by Storybook's esbuild configuration
// ------------------------------------------------------------
import {
  mockSveltekitStores
} from "./_node-chunks/chunk-WYXD273F.js";

// src/preset.ts
import { fileURLToPath } from "node:url";
import { withoutVitePlugins } from "@storybook/builder-vite";
import { viteFinal as svelteViteFinal } from "@storybook/svelte-vite/preset";

// src/plugins/config-overrides.ts
function configOverrides() {
  return {
    // SvelteKit sets SSR, we need it to be false when building
    name: "storybook:sveltekit-overrides",
    apply: "build",
    config: () => ({ build: { ssr: !1 } })
  };
}

// src/preset.ts
var core = {
  builder: import.meta.resolve("@storybook/builder-vite"),
  renderer: import.meta.resolve("@storybook/svelte/preset")
}, previewAnnotations = (entry = []) => [
  ...entry,
  fileURLToPath(import.meta.resolve("@storybook/sveltekit/preview"))
], viteFinal = async (config, options) => {
  let baseConfig = await svelteViteFinal(config, options);
  return {
    ...baseConfig,
    plugins: [
      // disable specific plugins that are not compatible with Storybook
      ...await withoutVitePlugins(baseConfig.plugins ?? [], [
        "vite-plugin-sveltekit-compile",
        "vite-plugin-sveltekit-guard"
      ]),
      configOverrides(),
      mockSveltekitStores()
    ]
  };
}, optimizeViteDeps = [
  "@storybook/sveltekit/internal/mocks/app/forms",
  "@storybook/sveltekit/internal/mocks/app/navigation",
  "@storybook/sveltekit/internal/mocks/app/stores",
  "@storybook/sveltekit/internal/mocks/app/state.svelte.js"
];
export {
  core,
  optimizeViteDeps,
  previewAnnotations,
  viteFinal
};
